src/ucp/core/ucp_worker.c

```diff
@@ -3078,13 +3059,82 @@ ucp_worker_keepalive_complete(ucp_worker_h worker, ucs_time_t now)
     worker->keepalive.round_count++;
 }
 
+static int ucp_worker_do_ep_keepalive(ucp_worker_h worker, ucs_time_t now)
+{
+    ucp_lane_index_t lane;
+    ucp_rsc_index_t rsc_index;
+    ucs_status_t status;
+    ucp_ep_h ep;
+
+    UCP_WORKER_THREAD_CS_CHECK_IS_BLOCKED(worker);
+
+    if (worker->keepalive.iter == &worker->all_eps) {
+        goto out_done;
+    }
+
+    ep = ucp_ep_from_ext_gen(ucs_container_of(worker->keepalive.iter,
+                                              ucp_ep_ext_gen_t, ep_list));
+
+    if ((ep->cfg_index == UCP_WORKER_CFG_INDEX_NULL) ||
+        (ep->flags & UCP_EP_FLAG_FAILED) ||
+        (ucp_ep_config(ep)->key.keepalive_lane == UCP_NULL_LANE)) {
+        goto out_done;
+    }
+
+    lane      = ucp_ep_config(ep)->key.keepalive_lane;
+    rsc_index = ucp_ep_get_rsc_index(ep, lane);
+
+    ucs_trace_func("worker %p: do keepalive on ep %p lane %u", worker, ep,
```

Comment:
plz remove, the trace on line 3097 below is enough
also it is not really a `trace_func`

Reply:
removed